### PR TITLE
Approvals at Branch Protection

### LINKED ISSUE #5251: Pull requests approvals limitations
Every repository could protected it's branches and set the approvals limitations before merge a pull request.

### PR DESCRIPTION
**Pull request by jonasfranz**
Fixes #5251 

This PR adds approvals limitations to branch protection. A pitfall is that it is currently only possible to assign Teams/Users with write permissions to branch protection. I've removed this for teams but not for users.

# Screenshots

|     |     |
|-------------|-------------|
| screenshot_2018-11-18 approvals 3         | screenshot_2018-11-18 approvals 2         |
|screenshot_2018-11-18 approvals 1|screenshot_2018-11-18 approvals|

### CODE REVIEW

File: models/branches.go
Comment (kolaente) [inline models/branches.go:114]: This will count multiple approvals by the same user as multiple approval reviews, wouldn't it?

File: models/branches.go
Comment (lunny) [inline models/branches.go]: why not just get all reviews where type is approval?

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@JonasFranzDEV any screenshot?

File: models/branches.go
Comment (jonasfranz) [reply models/branches.go]: I'm going to implement that together with group by to solve @kolaente s Problem.

**Comment by jonasfranz** [comment]
@lunny fixed your first comment and added screenshots.

File: models/review.go
Comment (jonasfranz) [reply models/review.go:169]: How should I change this to support mssql? I've no experience with mssql.

**Review by kolaente — APPROVED** [review]

**Comment by lunny** [comment]
right, blocked by #5132

File: models/review.go
Comment (lafriks) [reply models/review.go:169]: @JonasFranzDEV you can now reuse `GetReviewersByPullID` that will return list of per user latest reviews with type

**Comment by lafriks** [comment]
Do we allow merging if there is request for changes?

**Comment by lafriks** [comment]
Yes this can be added in other PR as an option

**Comment by lunny** [comment]
@lafriks I think @JonasFranzDEV is right, allow merging or not when there is request for changes should be another PR.

File: models/review.go
Comment (jonasfranz) [reply models/review.go:169]: I've used the function provided by #5132 instead.

**Comment by jonasfranz** [comment]
@lunny need your review

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
file conflict.

File: models/review.go
Comment (lunny) [reply models/review.go:169]: @lafriks @JonasFranzDEV I can confirm this is also not work on MSSQL.

**Comment by jonasfranz** [comment]
@lunny File conflicts resolved

**Comment by lunny** [comment]
Pull detail page will result in 500 on MSSQL.
image

**Comment by lunny** [comment]
Pull detail page will result in 500 on MSSQL.
image

**Comment by kolaente** [comment]
@lunny is this using the function from Review summaries?

**Comment by lunny** [comment]
Maybe this could be resolved by #5502

**Comment by jonasfranz** [comment]
Can you verify that? @lunny

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@JonasFranzDEV it seems it's has some bug on Mssql.  See
image

**Comment by ikostrubin-IP** [comment]
> @JonasFranzDEV it seems it's has some bug on Mssql. See
> image

Agree, the same error but with sqlite db.

**Comment by lunny** [comment]
You have to add `test1` to approvals list on protected branch. Otherwise his approvals will not be counted.

**Pull request merged**
